_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+1. Composite assignment operators, such as `+=`, will likely be provided automatically
+    in the future as long as the corresponding arithmetic/bitwise operator is available.
+    Allowing a direct implementation is currently not impossible anyway, because there are no references
+    to value types in the language.
+1. We are considering adding a C++-like
+    ["spaceship operator"](https://en.wikipedia.org/wiki/Three-way_comparison#High-level_languages),
+    i.e. `<=>`.
+    When defined, the compiler could use it to automatically generate all the other comparison operators.
+
+On the other hand, it is unlikely that we will allow redefining some other operators:
+
+1. `&&` and `||` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+    The additional complication is their short-circuiting behavior that may or may not be
+    appropriate, depending on the semantics of a particular UDVT.
+1. `=` is already provided by the compiler for all types and it is very unlikely that we will ever allow redefining it.
+
+#### Operators on Reference Types
+While UDVTs were the primary target, not all types can reasonably fit into a single slot.
+The way to solve this problem is to allow operators taking and returning structs.
+The main thing holding us back are the planned changes to reference types, in particular the fact
+that [data location](https://docs.soliditylang.org/en/latest/types.html#data-location) is going to
+become a part of the type itself.
+ To avoid having to introduce breaking changes later, we decided to defer support for structs until
+these changes are behind us.
+
+We want to eventually support operators on other reference types, but it is likely that this usage will
+require naming them, using some kind of yet-to-be-introduced user-defined reference type feature similar
+to UDVTs, or at least defining a type alias.
+
+#### Operators on Mixed Types
+Inability to define operators with arguments of different types is somewhat constraining and lifting this
+limitation is also a part of our vision for UDVTs.
+There are, however, multiple ways to achieve this, and explicitly defining an operator for every type combination
+is only one of them.
+Other possibilities involve generics specialized with traits or having implicit conversions.
+
+This part will unfortunately have to wait until we are done with the changes to the type system that are on our current roadmap.
+Doing them before would only result in something that would limit our design choices, and would have to be discarded
+in a breaking release.
+
+#### What Will Not Change?
+Due to concerns about the safety of the feature, there are some key elements that are very unlikely to ever change regarding operators.
+Operators will likely always be global and defined at file level with free functions.
+In terms of side-effects, they will stay `pure`.
```

Comment:
That's probably false.

Reply:
Oh, interesting. I know it's possible but I thought it's a very remote possibility, to the point that it would be reasonable to say this.

Also, non-pure operators do not seem to be useful enough to be worth considering.

In any case, removed that part.

Reply:
Once you have *any* operator on types in storage they're no longer pure, that's no exactly exotic :-).